source/MaterialXGenShader/Nodes/HwTexCoordNode.cpp

```diff
@@ -54,9 +54,25 @@ void TexCoordNodeGlsl::emitFunctionCall(const ShaderNode& node, GenContext& cont
         ShaderPort* texcoord = vertexData[variable];
         shadergen.emitLineBegin(stage);
         shadergen.emitOutput(node.getOutput(), true, false, context, stage);
-        shadergen.emitString(" = " + prefix + texcoord->getVariable(), stage);
+
+        // Extract the requested number of components from the texture coordinates (which may be a
```

Comment:
This will extract for texcoords, but as noted before if you can widen other streams like color then extraction is needed for those stream types as well. This only seems to be for color now, but might want to take into consideration other streams may have different channel counts in the future. (e.g. if  position-4 or tangent-4 is added)

Reply:
From what I understand, color nodes seem to already perform component extraction: see https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/source/MaterialXGenGlsl/Nodes/GeomColorNodeGlsl.cpp#L52. Is this correct?

Looking at the code though, I believe `GeomColorNode{Glsl,Msl}` should be merged into a `HwGeomColorNode`, as they are also almost identical, but perhaps this should be better done in a different PR as this does not seem to be linked with this issue.

Reply:
If you are worried that new nodes can now widen port types by mistake, it should be possible to add a boolean parameter to `addStageInput` such that the widening behavior is turned off by default (and an exception is thrown if the types do not match in that case). Would that help in that regard?

Reply:
I've now implemented the checks, as per the comment above, so there should be no risk of a node accidentally widening a port type by mistake. This also adds a check that was not there before, and should protect against requesting a type with a different type than the existing port type. This new behavior also documented in the header file.

I also took the liberty of applying the suggestions from the other PR (#1560), which were to use static strings for port names.

Reply:
Hi @madmann91 ,

- Thanks a lot for looking into this. 
- I'll take another look at this, but my main thought was do you always widen and trust that each type with perform an extraction from the widened type, or do you want an explicit "firewall" and as more types require widen the "firewall". 

- BTW, I have another query in  that there are derived code generators (such as the one in HDStorm), and wondering if those generators require any changes. Best if they do not but if so, we may want to make a note of that as a heads up.

Reply:
I believe that it's up to you to decide what is the expected behavior. From looking at the code, it seems that certain nodes are "polymorphic", in the sense that they can be created with different types (e.g. texcoord, ...), but others are not. Since this behavior can be specific to the backend or node type, I believe the current solution of explicitly requiring `shouldWiden = true` for polymorphic nodes is safe, as this makes sure the behavior remains the same as before, unless widening is requested: I thus believe that no code change is necessary anywhere else (at least if the current behavior is expected to stay the same). In any case, the new exception check should catch the case when nodes are incorrectly wired.